src/sync/mpsc/mod.rs

```diff
@@ -0,0 +1,693 @@
+
+        loop {
+            let mut state = decode_state(curr);
+
+            // The receiver end closed the channel.
+            if !state.is_open {
+                return None;
+            }
+
+            // This probably is never hit? Odds are the process will run out of
+            // memory first. It may be worth to return something else in this
+            // case?
+            assert!(state.num_messages < MAX_CAPACITY, "buffer space exhausted; sending this messages would overflow the state");
+
+            state.num_messages += 1;
+
+            // The channel is closed by all sender handles being dropped.
+            if close {
+                state.is_open = false;
+            }
+
+            let next = encode_state(&state);
+            let actual = self.inner.state.compare_and_swap(curr, next, Ordering::SeqCst);
+
+            if curr == actual {
+                // Block if the current number of pending messages has exceeded
+                // the configured buffer size
+                let park_self = match self.inner.buffer {
+                    Some(buffer) => state.num_messages > buffer,
+                    None => false,
+                };
+
+                // Only unpark the receive half if transitioning from 0 -> 1.
+                let unpark_recv = state.num_messages == 1;
+
+                return Some((park_self, unpark_recv));
+            }
+
+            curr = actual;
+        }
+    }
+
+    fn park(&mut self, can_park: bool) {
+        // TODO: clean up internal state if the task::park will fail
+
+        let task = if can_park {
+            Some(task::park())
+        } else {
+            None
+        };
+
+        self.maybe_parked = true;
+        *self.sender_task.lock().unwrap() = task;
+
+        // Send handle over queue
+        let t = self.sender_task.clone();
+        self.inner.parked_queue.push(t);
+    }
+}
+
+impl<T> Sink for Sender<T> {
+    type SinkItem = T;
+    type SinkError = SendError<T>;
+
+    fn start_send(&mut self, msg: T) -> StartSend<T, SendError<T>> {
+        self.start_send2(msg)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), SendError<T>> {
+        Ok(Async::Ready(()))
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        self.try_clone().expect("failed to clone sender; maximum number of senders already reached")
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        // Ordering between variables don't matter here
+        let prev = self.inner.num_senders.fetch_sub(1, Ordering::Relaxed);
+
+        if prev == 1 {
+            let _ = self.do_send(None, false);
+        }
+    }
+}
+
+/*
+ *
+ * ===== impl Receiver =====
+ *
+ */
+
+impl<T> Receiver<T> {
+    /// Closes the receiving half
+    ///
+    /// This prevents any further messages from being sent on the channel while
+    /// still enabling the receiver to drain messages that are buffered.
+    pub fn close(&mut self) {
```

Comment:
I think that we'll want to be sure to add this to other primitives as well to ensure a uniform API surface area.